# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
os.getcwd()

'C:\\Shalin\\Springboard bootcamp\\projects\\Guided Capstone project 1'

In [3]:
os.listdir()

['.ipynb_checkpoints',
 'data',
 'figures',
 'Guided Capstone Project Report.docx',
 'GuidedCapstoneStep3HL.ipynb',
 'GuidedCapstoneStep4 and step5HL.ipynb',
 'GuidedCapstoneStep6HL.ipynb',
 'GuidedCapstone_Step2HL.ipynb',
 'models',
 'Step 1 -Problem Identification',
 'Step 2 - Data Wrangling',
 'Step 3 - EDA',
 'Step 4 - Data preprocessing and training data development',
 'Step 5 - Modeling',
 'Step 6 - Documentation',
 'updated_ski_data.csv',
 '~$ided Capstone Project Report.docx']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [4]:
filename = 'C:\\Shalin\\Springboard bootcamp\\projects\\Guided Capstone project 1\\data\\step3_output.csv'
df = pd.read_csv(filename)

df.head(5)

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  LongestRun_mi  \
0        0.0          0          2     2  ...            1.0   
1        0.0          0          0     0  ...            2.0   
2        0.0          0          0     0  ...            1.0   
3        0.0          1          0     2  ...            2.0   
4        0.0          0          1     2  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
df.dtypes

Name                  object
state                 object
summit_elev            int64
vertical_drop          int64
base_elev              int64
trams                  int64
fastEight            float64
fastSixes              int64
fastQuads              int64
quad                   int64
triple                 int64
double                 int64
surface                int64
total_chairs           int64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
dtype: object

In [6]:
dfo=df.select_dtypes(include=['object']) # select object type columns
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

In [7]:
df.columns

Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       ...
       'state_Rhode Island', 'state_South Dakota', 'state_Tennessee',
       'state_Utah', 'state_Vermont', 'state_Virginia', 'state_Washington',
       'state_West Virginia', 'state_Wisconsin', 'state_Wyoming'],
      dtype='object', length=388)

In [8]:
df.shape

(330, 388)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:
df = df.loc[:,~df.columns.str.startswith('Name')]
df.columns

Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'state_Alaska', 'state_Arizona', 'state_California',
       'state_Colorado', 'state_Connecticut', 'state_Idaho', 'state_Illinois',
       'state_Indiana', 'state_Iowa', 'state_Maine', 'state_Maryland',
       'state_Massachusetts', 'state_Michigan', 'state_Minnesota',
       'state_Missouri', 'state_Montana', 'state_Nevada',
       'state_New Hampshire', 'state_New Jersey', 'state_New Mexico',
       'state_New York', 'state_North Carolina', 'state_Ohio', 'state_Oregon',
       'state_Pennsylvania', 'state_Rhode Island', 'state_South Dakota',
       'state_Tennessee', 'state_Utah', 'state_Vermont', 

In [10]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['AdultWeekend'], axis = 1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

C:\Users\gosal\Anaconda3_2\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\gosal\Anaconda3_2\lib\site-packages\ipykernel_launcher.py:14: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  


In [11]:
X.shape

(330, 58)

In [12]:
y.shape

(330,)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [13]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [14]:
X_train.shape

(247, 58)

In [15]:
X_test.shape

(83, 58)

In [16]:
y_train.shape

(247,)

In [17]:
y_test.shape

(83,)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [18]:
#Linear Regresion model
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [19]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [20]:
y_pred.mean()

65.35043499401165

## Review Model Outcomes — Iterate over additional models as needed

In [21]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
evs = explained_variance_score(y_test, y_pred)
evs

0.9238399289235384

In [22]:
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test, y_pred)
mae

5.458186271009807

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [23]:
model.intercept_

11.528620020106317

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [24]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by = ['Coefficient'], ascending = False)

Coefficient
state_Tennessee         19.335288
state_North Carolina    12.890576
state_West Virginia     10.473194
state_Montana           10.262626
state_Maryland          10.214804
state_Massachusetts      9.427458
state_New Mexico         9.285782
state_South Dakota       8.400387
state_Idaho              7.341540
state_California         7.308286
state_Virginia           6.984581
state_New Hampshire      6.495548
state_Oregon             6.110581
state_Wyoming            6.089918
state_Ohio               6.057991
state_Colorado           5.628094
state_Indiana            5.323363
state_Arizona            5.080908
state_Connecticut        4.392514
state_Washington         3.766343
state_Rhode Island       3.546624
state_Pennsylvania       3.539639
state_Minnesota          3.254610
state_Iowa               3.203804
state_Alaska             2.991617
state_Missouri           2.648334
state_Vermont            2.130195
state_Maine              2.106806
state_Utah               1.861217
state_New Jersey         1.549409
state_Illinois           1.302129
fastEight                1.199838
state_Nevada             0.960033
fastSixes                0.876954
state_New York           0.860763
state_Wisconsin          0.853255
surface                  0.850086
AdultWeekday             0.805449
triple                   0.767123
state_Michigan           0.527350
quad                     0.497814
LongestRun_mi            0.494492
total_chairs             0.434673
trams                    0.361827
TerrainParks             0.231505
double                   0.085488
fastQuads                0.026724
Runs                     0.018971
daysOpenLastYear         0.012344
projectedDaysOpen        0.008306
NightSkiing_ac           0.006971
vertical_drop            0.005602
base_elev                0.005314
Snow Making_ac           0.005302
summit_elev              0.004701
yearsOpen                0.004272
averageSnowfall          0.002873
SkiableTerrain_ac        0.000081

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [25]:
# Rebuid model without the 'state' columns
X_train_2= X_train.loc[:,~X_train.columns.str.startswith('state')]
X_test_2= X_test.loc[:,~X_test.columns.str.startswith('state')]
X_train_2.shape

(247, 23)

In [26]:
X_test_2.shape

(83, 23)

In [27]:
y_train.shape

(247,)

In [28]:
y_test.shape

(83,)

In [29]:
#Apply Linear Regresion model on the updated dataframe without state columns
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model_2 = lm.fit(X_train_2,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [30]:
# Make a variable called y_pred_2 and assign it the result of calling predict() on our model variable with parameter X_test_2
y_pred_2 = lm.predict(X_test_2)
y_pred_2

array([ 57.14616419,  86.8599285 ,  82.03177525,  78.39042297,
        66.5744501 ,  78.08307995,  37.76460325,  46.60603034,
        88.77568606,  81.95695713,  53.29838454,  79.00291795,
        53.96834281,  65.61008556,  64.66559007,  89.83993959,
        38.08117798,  76.78396063,  64.00664539,  43.03858462,
       107.4879086 ,  37.26044502,  52.45963558,  50.43560051,
        50.38998591,  64.85225674,  66.50830831,  79.28746149,
        63.88403434,  63.79504802,  44.58120304,  52.56194247,
        78.37826529,  56.62723725,  75.50333554,  54.64876848,
        22.81884559,  27.82246864,  55.89503558,  36.37623649,
        47.03352578,  64.64293288,  88.91323242,  66.46259944,
        69.69997058,  53.64652702,  74.0126918 , 168.0269444 ,
        30.7155824 ,  61.25671249,  63.66956588,  60.77015884,
        54.56128682,  48.65721145,  60.81966617,  42.58978672,
        86.39933278,  28.45559259, 116.64443575, 116.74917128,
        57.75642506,  51.27178403,  40.09807659,  86.36

## Review Model Outcomes — Iterate over additional models as needed

In [31]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
evs_2 = explained_variance_score(y_test, y_pred_2)
evs_2

0.9129263010032647

In [32]:
from sklearn.metrics import mean_absolute_error
mae_2 = mean_absolute_error(y_test, y_pred_2)
mae_2

5.774676977603139

Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [33]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X_train_2.columns, columns=['Coefficient']).sort_values(by = ['Coefficient'], ascending = False)

Coefficient
quad                  0.985113
AdultWeekday          0.843314
fastEight             0.840020
triple                0.739080
surface               0.715392
trams                 0.608807
fastSixes             0.444595
LongestRun_mi         0.325382
TerrainParks          0.262260
total_chairs          0.250800
fastQuads             0.235895
Runs                  0.025746
projectedDaysOpen     0.021540
averageSnowfall       0.016872
double                0.010336
vertical_drop         0.005307
NightSkiing_ac        0.004588
summit_elev           0.003909
base_elev             0.003721
daysOpenLastYear      0.003607
Snow Making_ac        0.003402
yearsOpen             0.002667
SkiableTerrain_ac     0.000161

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [34]:
# Rebuid model without the 'summit_elev' and 'base_elev' columns
X_train_3 = X_train_2.drop(columns = ['summit_elev','base_elev'], axis = 1)
#X_train_3= X_train_2.loc[:,~X_train_2.columns.str.startswith('summit_')]
X_test_3 = X_test_2.drop(columns = ['summit_elev','base_elev'], axis = 1)
#X_train_3.columns

In [35]:
X_train_3.shape

(247, 21)

In [36]:
X_test_3.shape

(83, 21)

In [37]:
#Apply Linear Regresion model on the updated dataframe without state and summit_elev columns
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model_3 = lm.fit(X_train_3,y_train)

In [38]:
# Make a variable called y_pred_3 and assign it the result of calling predict() on our model variable with parameter X_test_3
y_pred_3 = lm.predict(X_test_3)
y_pred_3

array([ 57.56092308,  86.22270561,  82.99485852,  78.54960299,
        65.61898214,  78.78061537,  37.63571693,  46.47633066,
        88.60261167,  81.61560711,  53.00445968,  80.28567299,
        54.27958056,  65.37766377,  64.31281473,  89.30078156,
        41.54721403,  77.54936094,  65.58177026,  42.97767217,
       107.19626015,  37.18107109,  51.83368793,  50.21971934,
        50.37708102,  64.69389954,  66.66871949,  75.6743338 ,
        64.66343211,  63.82418106,  44.67867556,  52.42213291,
        78.77586227,  56.16024062,  75.25707232,  54.55478491,
        22.49297277,  27.67449697,  55.65778594,  35.90669865,
        47.43611923,  64.96991121,  88.83200923,  66.157902  ,
        69.84300334,  52.95571493,  72.66937928, 168.05976665,
        30.58911574,  61.9763069 ,  63.22094397,  61.67197364,
        54.49836855,  48.38431729,  60.92033123,  43.05883152,
        85.61022989,  28.98120957, 115.72838335, 116.18421557,
        57.47037211,  51.50332416,  39.76958498,  85.88

## Review Model Outcomes — Iterate over additional models as needed

In [39]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
evs_3 = explained_variance_score(y_test, y_pred_3)
evs_3

0.9178131444563123

In [40]:
from sklearn.metrics import mean_absolute_error
mae_3 = mean_absolute_error(y_test, y_pred_3)
mae_3

5.704383942571344

Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [41]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X_train_3.columns, columns=['Coefficient']).sort_values(by = ['Coefficient'], ascending = False)

Coefficient
fastEight             1.666805
surface               1.026291
trams                 0.944729
AdultWeekday          0.841002
quad                  0.696394
total_chairs          0.585870
triple                0.400869
LongestRun_mi         0.319678
double                0.308554
TerrainParks          0.211647
fastSixes             0.160117
fastQuads             0.058740
Runs                  0.030664
projectedDaysOpen     0.022681
averageSnowfall       0.018511
NightSkiing_ac        0.005926
daysOpenLastYear      0.003053
Snow Making_ac        0.002809
yearsOpen             0.002732
vertical_drop         0.001478
SkiableTerrain_ac     0.000203

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

Model Selection:

| Model            | Explained Variance| Mean Absolute Error|Features Dropped                 |
| -----------------|-------------------|--------------------|---------------------------------|                    
| Model 1.         | 0.92              | 5.46               | -                               |
| Model 2.         | 0.91              | 5.77               |'state'                          |
| Model 3.         | 0.92              | 5.70               |'state','summit_elev','base_elev |

**Conclusion**
Model 1 is the best model since due to highest explained variance and lowest mean absolute error. Considering different characteristics and vairables that apply to each resort, it is important for Big Mountain management to be able to stick to a model that is able to predict with higest confidence what variables/ factors would drive ski resort lift ticket prices since ticket pricing in order to increase the revenue and offset for the increasing CAPEX it incurred due to addition of a new chair lift device, and at the same time, low error rate. If you note, all 3 models are almost leading to same % confidence and the error is also somewhat in the same range. But going by the metrics as indicated in above table, Model 1 is best